### ISSUE TITLE
[Bug] Possible problem with handling of connection

### ISSUE THREAD

**Issue by brianleect** [brianleect/binance-pump-alerts#13]
Thanks Brian. I am currently testing and eventually gets stuck on the following message:

_"DUMP: TCTUSDT / Change: -1.16 % / Price: 0.023277 Interval: 1m
Time taken to extract and append: 6.69248366355896
Extracting after 15 s"_

And I have to interrupt the process and then restart it and have it start up again. I think maybe it is due to some failures in my internet connection, but I think the code should try to reconnect after some specific period of time, to avoid getting stuck at that point. Update: This situation occurred three times during the day. 

Another bug occurs when too many notifications accumulate for Telegram message and I get the following error:

_"DUMP: VTHOUSDT / Change: -1.09 % / Price: 0.005849 Interval: 1m
DUMP: STORJUSDT / Change: -1.15 % / Price: 0.9025 Interval: 1m
DUMP: MANAUSDT / Change: -1.03 % / Price: 0.6822 Interval: 1m
Telegram bot error
Telegram bot error
Telegram bot error
Telegram bot error
Telegram bot error
DUMP: BONDUSDT / Change: -2.26 % / Price: 28.76 Interval: 3m
DUMP: VTHOUSDT / Change: -1.09 % / Price: 0.005849 Interval: 1m"_

I don't know if configuring a minimum pause between each notification can fix it.

_Originally posted by @Juanjacinto123 in https://github.com/brianleect/binance-pump-alerts/issues/10#issuecomment-878443284_

**Comment by brianleect** [comment]
Thanks for raising the issue on the error @Juanjacinto123 ,  I have created a separate issue to prevent clogging up the original thread regarding the volume feature.

Firstly, regarding connection issues when you interrupted the message may I know if it stated which line of code it was stuck at? I took a look at the code used for extraction currently and it "auto-reconnecting" should have already been taken into account? 

![image](https://user-images.githubusercontent.com/63389110/125498871-b50e6f6e-f0e8-4519-a02a-80d31066fb3b.png)

^ Code referenced. To further clarify on this issue, did this occur with the previous version without variable extraction as well? Or is it only after I made the current update. In theory it shouldn't have made a difference though as there were no breaking changes, just minor modifications.

Secondly, regarding the "telegram bot" error, it can be mostly ignored as if I'm not mistaken it only occurs while multiple telegram messages is trying to be sent within a short timeframe. I have previously programmed it such that it re-attempts to send the message again if it fails after sleeping for 0.5s.

![image](https://user-images.githubusercontent.com/63389110/125497335-c60967c6-bf02-49a4-943c-89564993a553.png)

^ Can be observed from the code above. I had it print "Telegram bot error" as a debug feature. All triggered alerts should still be sent as per normal just that due to large amount of notifications being sent at once it needs a short period before being able to send the next alert as a telegram message.

**Comment by brianleect** [comment]
> A few minutes ago I received the following error:
> 
> Error: HTTPSConnectionPool(host='api.binance.com', port=443):
> 
> _"PUMP: ACMUSDT / Change: 3.28 /% Price: 6.765 Interval: 10m PUMP: SLPUSDT / Change: 4.28 /% Price: 0.2918 Interval: 30m Time taken to extract and append: 5.10305118560791 Extracting after 60 s Time taken to extract and append: 2.422328472137451 Extracting after 60 s Time taken to extract and append: 6.2276291847229 Extracting after 60 s Time taken to extract and append: 2.349924325942993 Extracting after 60 s Error: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ticker/price (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002100F3D63C8>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')) Retrying in 1s Time taken to extract and append: 15.77428674697876 Extracting after 60 s Time taken to extract and append: 6.378751993179321 Extracting after 60 s"_
> 
> This is the setup I am using:
> 
> _intervals = ['1m', '3m', '5m', '10m', '15m', '30m']
> outlier_param = {'1m':0.015,'3m':0.02,'5m':0.03,'10m':0.03,'15m':0.03,'30m':0.04}
> pairs_of_interest = ['USDT'] # Other options include 'BTC' , 'ETH'
> watchlist = [] # E.g. ['ADAUSDT', 'ETHUSDT'] # Note that if watchlist has pairs, ONLY pairs in watchlist will be monitored
> 
> FUTURE_ENABLED=False # Determine whether to look at future markets
> DUMP_ENABLED = True # Determine whether to look at DUMP
> MIN_ALERT_INTERVAL = '90s' # Minimum interval between alerts for SAME pair
> RESET_INTERVAL = '3h' # Interval for clearing array to prevent MEM ERROR
> PRINT_DEBUG = True # If false we do not print messages
> EXTRACT_INTERVAL = '60s' # Minimum interval is 2s but reccommend for 5s and above due to 10054 errors_
> 
> Update: looking at the operation log I noticed that this error has occurred twice during the day.

Regarding the error you mentioned, it seemed to also have been faced in the thread below

https://github.com/sammchardy/python-binance/issues/227

![image](https://user-images.githubusercontent.com/63389110/125500439-9df09aa5-f98e-42a0-a67b-bd9cb1c75894.png)

From the getPrice code the try except loop portion did not seem to have been triggered at all as well. Seeing that none of the error messages "Error: __ " and "Retrying in 1s" got printed to console.

![image](https://user-images.githubusercontent.com/63389110/125501333-80741ba2-4e20-45a3-9014-c5bcdef7cab5.png)
^ Error messages can be seen above

As previously stated in the response above, do update me if this was an error that occurred only after my update and if possible which line of code the program was stuck at which should be outputted when you forcefully interrupted the program so I can better debug as well. 

A possible solution might be increasing the sleep duration in the except loop portion of the code, however I have doubts on it working considering the error messages were not even printed at all.

**Comment by Juanjacinto123** [comment]
> Thanks for raising the issue on the error @Juanjacinto123 , I have created a separate issue to prevent clogging up the original thread regarding the volume feature.
> 
> Firstly, regarding connection issues when you interrupted the message may I know if it stated which line of code it was stuck at? I took a look at the code used for extraction currently and it "auto-reconnecting" should have already been taken into account?
> 

It got stuck right after this message:

_Extracting after 15 s_

> ![image](https://user-images.githubusercontent.com/63389110/125498871-b50e6f6e-f0e8-4519-a02a-80d31066fb3b.png)
> 
> ^ Code referenced. To further clarify on this issue, did this occur with the previous version without variable extraction as well? Or is it only after I made the current update. In theory it shouldn't have made a difference though as there were no breaking changes, just minor modifications.
> 

It happened when I started using the extraction interval of 15s and 60s, but I was testing it on my computer and my Internet connection has been failing lately, maybe that could have generated this error. 

The Original version (with 1s extraction interval) I have always tested it on my VPS, maybe for that reason on the vps it has not failed, because the internet connection is more stable.

I could try this new version with 60s extraction interval on the VPS to see how it works. 

> Secondly, regarding the "telegram bot" error, it can be mostly ignored as if I'm not mistaken it only occurs while multiple telegram messages is trying to be sent within a short timeframe. I have previously programmed it such that it re-attempts to send the message again if it fails after sleeping for 0.5s.
> 
> ![image](https://user-images.githubusercontent.com/63389110/125497335-c60967c6-bf02-49a4-943c-89564993a553.png)
> 
> ^ Can be observed from the code above. I had it print "Telegram bot error" as a debug feature. All triggered alerts should still be sent as per normal just that due to large amount of notifications being sent at once it needs a short period before being able to send the next alert as a telegram message.
> 
> ** Perhaps for clarity I should rename the the print message when a telegram message has failed to be sent with "Attempting to resend telegram message in X seconds"

**Comment by Juanjacinto123** [comment]
> > A few minutes ago I received the following error:
> > Error: HTTPSConnectionPool(host='api.binance.com', port=443):
> > _"PUMP: ACMUSDT / Change: 3.28 /% Price: 6.765 Interval: 10m PUMP: SLPUSDT / Change: 4.28 /% Price: 0.2918 Interval: 30m Time taken to extract and append: 5.10305118560791 Extracting after 60 s Time taken to extract and append: 2.422328472137451 Extracting after 60 s Time taken to extract and append: 6.2276291847229 Extracting after 60 s Time taken to extract and append: 2.349924325942993 Extracting after 60 s Error: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ticker/price (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002100F3D63C8>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')) Retrying in 1s Time taken to extract and append: 15.77428674697876 Extracting after 60 s Time taken to extract and append: 6.378751993179321 Extracting after 60 s"_
> > This is the setup I am using:
> > _intervals = ['1m', '3m', '5m', '10m', '15m', '30m']
> > outlier_param = {'1m':0.015,'3m':0.02,'5m':0.03,'10m':0.03,'15m':0.03,'30m':0.04}
> > pairs_of_interest = ['USDT'] # Other options include 'BTC' , 'ETH'
> > watchlist = [] # E.g. ['ADAUSDT', 'ETHUSDT'] # Note that if watchlist has pairs, ONLY pairs in watchlist will be monitored
> > FUTURE_ENABLED=False # Determine whether to look at future markets
> > DUMP_ENABLED = True # Determine whether to look at DUMP
> > MIN_ALERT_INTERVAL = '90s' # Minimum interval between alerts for SAME pair
> > RESET_INTERVAL = '3h' # Interval for clearing array to prevent MEM ERROR
> > PRINT_DEBUG = True # If false we do not print messages
> > EXTRACT_INTERVAL = '60s' # Minimum interval is 2s but reccommend for 5s and above due to 10054 errors_
> > Update: looking at the operation log I noticed that this error has occurred twice during the day.
> 
> Regarding the error you mentioned, it seemed to also have been faced in the thread below
> 
> [sammchardy/python-binance#227](https://github.com/sammchardy/python-binance/issues/227)
> 
> ![image](https://user-images.githubusercontent.com/63389110/125500439-9df09aa5-f98e-42a0-a67b-bd9cb1c75894.png)
> 
> From the getPrice code the try except loop portion did not seem to have been triggered at all as well. Seeing that none of the error messages "Error: __ " and "Retrying in 1s" got printed to console.
> 
> ![image](https://user-images.githubusercontent.com/63389110/125501333-80741ba2-4e20-45a3-9014-c5bcdef7cab5.png)
> ^ Error messages can be seen above
> 
> As previously stated in the response above, do update me if this was an error that occurred only after my update and if possible which line of code the program was stuck at which should be outputted when you forcefully interrupted the program so I can better debug as well.
> 
> A possible solution might be increasing the sleep duration in the except loop portion of the code, however I have doubts on it working considering the error messages were not even printed at all.

Indeed the error messages were printed in the console, here you can see them:

_"Error: HTTPSConnectionPool(host='api.binance.com', port=443): Max retries exceeded with url: /api/v3/ticker/price (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002100F3D63C8>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')) 
Retrying in 1s"_

**Comment by brianleect** [comment]
@Juanjacinto123 , you could take a look at the newest updated version. I have added the parameters to allow you to easily modify re-attempt intervals for both sending of telegram messages and getting price data. Currently both are set to a default of '1s'.

Ah my bad, didn't notice the portion which printed out my error messages. Thanks for pointing it out. You could consider adjusting the parameter "GET_PRICE_FAIL_INTERVAL" to a larger value maybe '10s'? '15s'? However, suspect that it is likely due to internet instability you mentioned which might not be resolved by this method. It is still worth taking a look at though.

Before trying the new version on the VPS, you could consider saving a copy of the old version so that if that my update was the cause of the issue you would still having a working version in the meantime while I fix it.

I'll be heading to sleep now. If it still doesn't work for you after adjusting the parameters or on your VPS, I'll try running the code for an extended period of time on my machine as well and see if I can replicate this issue.

**Comment by Juanjacinto123** [comment]
> @Juanjacinto123 , you could take a look at the newest updated version. I have added the parameters to allow you to easily modify re-attempt intervals for both sending of telegram messages and getting price data. Currently both are set to a default of '1s'.
> 
> Ah my bad, didn't notice the portion which printed out my error messages. Thanks for pointing it out. You could consider adjusting the parameter "GET_PRICE_FAIL_INTERVAL" to a larger value maybe '10s'? '15s'? However, suspect that it is likely due to internet instability you mentioned which might not be resolved by this method. It is still worth taking a look at though.
> 
> Before trying the new version on the VPS, you could consider saving a copy of the old version so that if that my update was the cause of the issue you would still having a working version in the meantime while I fix it.
> 
> I'll be heading to sleep now and will run be testing the current version on my computer overnight. If it still doesn't work for you after adjusting the parameters or on your VPS do give me an update.

Ok thank very much for your time. I will let you know.

**Comment by brianleect** [comment]
@Juanjacinto123 , just to update on the current status of running the new version of the script on my PC. It has ran close to 20 hours consecutively without any errors raised for me. I did not adjust any parameters as well.

**Comment by brianleect** [comment]
@Juanjacinto123 , any updates? If everything is ok I'll be closing the issue.

**Comment by Juanjacinto123** [comment]
> @Juanjacinto123 , any updates? If everything is ok I'll be closing the issue.

Yes everything is working fine!
